### PR TITLE
Expose more data on PositionUtil for easier quote generation

### PR DESCRIPTION
**Pull request by odcheung**
- Add WhirlpoolData, TickArrayData to Position (client object) to allow for an easier time to generate quotes
- Change collect fee to user PositionData instead of client Position since it doesn't need that much info
- Add new TickUtil.getOffsetIndex to calculate array offset index for a position's tick

### Testing
- The `client.getPositions` is used by Orca UI
- Confirmed that it successfully fetches all positions and have identical data to live (test case had 7 positions, SOL & nonSOL pairs)
- The UX performance is actually faster because the old version does not do batch RPC

https://app.asana.com/0/0/1203431351729803/f

### CODE REVIEW

**Review by yugure-orca — CHANGES_REQUESTED** [review]

File: sdk/src/impl/whirlpool-client-impl.ts [hunk 1]
```diff
@@ -96,23 +98,77 @@ export class WhirlpoolClientImpl implements WhirlpoolClient {
...
+        }
+      })
+    );
+    await this.ctx.fetcher.listTickArrays(Array.from(tickArrayAddresses), true);
 
+    // Use getPosition and the prefetched values to generate the Positions
+    const results = await Promise.all(
+      positionAddresses.map(async (pos) => {
+        try {
+          const position = await this.getPosition(pos, false);
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/impl/whirlpool-client-impl.ts:165]: ah nice. good catch.

**Comment by odcheung** [comment]
@yugure-orca addressed your comments!

File: sdk/src/impl/whirlpool-client-impl.ts [hunk 2]
```diff
@@ -96,23 +97,77 @@ export class WhirlpoolClientImpl implements WhirlpoolClient {
...
-        return [address, null];
-      }
-
-      return [address, new PositionImpl(this.ctx, AddressUtil.toPubKey(address), positionAccount)];
-    });
+    // TODO: Prefetch and use fetcher as a cache - Think of a cleaner way to prefetch
+    const positions = await this.ctx.fetcher.listPositions(positionAddresses, refresh);
+    const whirlpoolAddr = positions
+      .map((position) => position?.whirlpool.toBase58())
+      .flatMap((x) => (!!x ? x : []));
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/impl/whirlpool-client-impl.ts:132]: if you use filter, typescript wouldn't be able to remove the 'null' part from the 'PublicKey | null' type. Here we are telling typescript that we are either returning something or nothing at all, rendering the type back to PublicKey.

**Review by yugure-orca — APPROVED** [review]
LGTM!

**Pull request merged**